rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -1324,4 +1457,146 @@ size_t Player::get_number_of_registered_on_play_msg_post_callbacks()
+    ts_handle->data);
+}
+
+bool PlayerImpl::PlayerClient::is_include_request_message(const rclcpp::SerializedMessage & message)
+{
+  auto type = get_msg_event_type(message);
+
+  // Ignore response message
+  if (type == service_msgs::msg::ServiceEventInfo::RESPONSE_SENT ||
+    type == service_msgs::msg::ServiceEventInfo::RESPONSE_RECEIVED)
+  {
+    return false;
+  }
+
+  // Ignore metadata message
+  if (type == service_msgs::msg::ServiceEventInfo::REQUEST_SENT &&
+    client_side_type_ == introspection_type::METADATA)
+  {
+    return false;
+  }
+  if (type == service_msgs::msg::ServiceEventInfo::REQUEST_RECEIVED &&
+    service_side_type_ == introspection_type::METADATA)
+  {
+    return false;
+  }
+
+  if (client_side_type_ == introspection_type::UNKNOW &&
+    type == service_msgs::msg::ServiceEventInfo::REQUEST_SENT)
+  {
+    if (message.size() <= rosbag2_cpp::get_serialization_size_for_service_metadata_event()) {
+      client_side_type_ = introspection_type::METADATA;
+      RCLCPP_WARN(
+        logger_,
+        "The configuration of introspection for '%s' client is metadata !",
+        service_name_.c_str());
+      return false;
+    } else {
+      client_side_type_ = introspection_type::CONTENTS;
+    }
+  }
+  if (service_side_type_ == introspection_type::UNKNOW &&
+    type == service_msgs::msg::ServiceEventInfo::REQUEST_RECEIVED)
+  {
+    if (message.size() <= rosbag2_cpp::get_serialization_size_for_service_metadata_event()) {
+      service_side_type_ = introspection_type::METADATA;
+      RCLCPP_WARN(
+        logger_,
+        "The configuration of introspection for '%s' service is metadata !",
+        service_name_.c_str());
+      return false;
+    } else {
+      service_side_type_ = introspection_type::CONTENTS;
+    }
+  }
+
+  // If there are request send info and request receive info, only send request send info.
+  if (client_side_type_ == introspection_type::CONTENTS &&
+    service_side_type_ == introspection_type::CONTENTS &&
+    type == service_msgs::msg::ServiceEventInfo::REQUEST_RECEIVED)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+void PlayerImpl::PlayerClient::async_send_request(const rclcpp::SerializedMessage & message)
+{
+  void * ros_message = new uint8_t[message_members_->size_of_];
+  message_members_->init_function(
+    ros_message, rosidl_runtime_cpp::MessageInitialization::ZERO);
+
+  int ret = rmw_deserialize(&message.get_rcl_serialized_message(), ts_, ros_message);
+  if (ret == RMW_RET_OK) {
+    if (client_->service_is_ready()) {
+      // members_[0]: info, members_[1]: request, members_[2]: response
+      auto request_offset = message_members_->members_[1].offset_;
+      auto request_addr = reinterpret_cast<size_t>(ros_message) + request_offset;
+      client_->async_send_request(
+        reinterpret_cast<void *>(*reinterpret_cast<size_t *>(request_addr)));
+    } else {
+      RCLCPP_ERROR(
+        logger_,
+        "'%s' service isn't ready !",
```

Comment:
Comments from MichaelOrlov

```suggestion
        logger_, "Service request hasn't been sent. The '%s' service isn't ready !",
```

Reply:
Updated